File: biophysics/VClamp.cpp
// VClamp.cpp --- 
// 
// Filename: VClamp.cpp
// Description: 
// Author: 
// Maintainer: 
// Created: Fri Feb  1 19:30:45 2013 (+0530)
// Version: 
// Last-Updated: Wed Feb  6 12:06:18 2013 (+0530)
//           By: subha
//     Update #: 268
// URL: 
// Keywords: 
// Compatibility: 
// 
// 

// Commentary: 
// 
// Implementation of Voltage Clamp
// 
// 

// Change log:
// 
// 
// 
// 
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as
// published by the Free Software Foundation; either version 3, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program; see the file COPYING.  If not, write to
// the Free Software Foundation, Inc., 51 Franklin Street, Fifth
// Floor, Boston, MA 02110-1301, USA.
// 
// 

// Code:

#include "header.h"
#include "Dinfo.h"
#include "VClamp.h"

using namespace moose;

const unsigned int VClamp::DERIVATIVE_ON_PV = 1;
const unsigned int VClamp::PROPORTIONAL_ON_PV = 2;


SrcFinfo1< double >* VClamp::currentOut()
{
    static SrcFinfo1< double > currentOut("currentOut",
                                          "Sends out current output of the clamping circuit. This should be"
                                          " connected to the `injectMsg` field of a compartment to voltage clamp"
                                          " it.");
    return &currentOut;
}


const Cinfo * VClamp::initCinfo()
{
    static DestFinfo process("process",
                             "Handles 'process' call on each time step.",
                             new ProcOpFunc< VClamp >( &VClamp::process));
    static DestFinfo reinit("reinit",
                            "Handles 'reinit' call",
                            new ProcOpFunc< VClamp >( &VClamp::reinit));
    static Finfo * processShared[] = {
        &process,
        &reinit
    };

    static SharedFinfo proc("proc",
                            "Shared message to receive Process messages from the scheduler",
                            processShared, sizeof(processShared) / sizeof(Finfo*));

    static ValueFinfo< VClamp, double> command("command",
                                                        "Command input of the clamp circuit.",
                                                        &VClamp::setCommand,
                                                        &VClamp::getCommand);
    static ValueFinfo< VClamp, unsigned int> mode("mode",
                                                  "Working mode of the PID controller."
                                                  "\nmode = 0, standard PID with proportional, integral and derivative"
                                                  "\n\tall acting on the error."
                                                  "\nmode = 1, derivative action based on command input"
                                                  "\nmode = 2, proportional action and derivative action are based on"
                                                  "\ncommand input.",
                                                  &VClamp::setMode,
                                                  &VClamp::getMode);
    static ValueFinfo< VClamp, double> ti("ti",
                                          "Integration time of the PID controller. Defaults to 1e9, i.e. integral"
                                          "\naction is negligibly small.",
                                          &VClamp::setTi,
                                          &VClamp::getTi);
    
    static ValueFinfo< VClamp, double> td("td",
                                          "Derivative time of the PID controller. This defaults to 0,"
                                          "\ni.e. derivative action is unused.",
                                          &VClamp::setTd,
                                          &VClamp::getTd);
    static ValueFinfo< VClamp, double> tau("tau",
                                          "Time constant of the lowpass filter at input of the PID"
                                           "\ncontroller. This smooths out abrupt changes in the input. Set it to "
                                           "\n5 * dt or more to avoid overshoots.",
                                          &VClamp::setTau,
                                          &VClamp::getTau);
    static ValueFinfo< VClamp, double> gain("gain",
                                          "Proportional gain of the PID controller.",
                                          &VClamp::setGain,
                                          &VClamp::getGain);    
    static ReadOnlyValueFinfo< VClamp, double> current("current",
                                            "The amount of current injected by the clamp into the membrane.",
                                            &VClamp::getCurrent);


    static DestFinfo sensed("sensed",
                              "Handles membrane potential read from compartment. The `VmOut` message"
                               " of the Compartment object should be connected to this.",
                                new OpFunc1< VClamp, double>( &VClamp::setVin));

    static Finfo* vclampFinfos[] = {
        currentOut(),
        &command,
        &current,
        &sensed,
        &mode,
        &ti,
        &td,
        &tau,
        &gain,
        &proc
    };

    static string doc[] = {
        "Name", "VClamp",
        "Author", "Subhasis Ray",
        "Description", "Voltage clamp object for holding neuronal compartments at a specific"
        " voltage. This implementation uses a builtin RC circuit to filter the"
        "\ncommand input and then use a PID to bring the sensed voltage (Vm from"
        "\ncompartment) to the filtered command potential."
        "\n"
        "\nUsage: Connect the `currentOut` source of VClamp to `injectMsg`"
        "\ndest of Compartment. Connect the `VmOut` source of Compartment to"
        "\n`sensed` dest of VClamp. Either set `command` field to a"
        "\nfixed value, or connect an appropriate source of command potential"
        "\n(like the `outputOut` message of an appropriately configured"
        "\nPulseGen) to `set_command` dest."
        "\n The default settings for the RC filter and PID controller should be"
        "\nfine. For step change in command voltage, good defaults with"
        "integration time step dt are as follows:"
        "\ntime constant of RC filter, tau = 5 * dt"
        "\nproportional gain of PID, gain = Cm/dt where Cm is the membrane"
        "\n\tcapacitance of the compartment"
        "\nintegration time of PID, ti = dt"
        "\nderivative time  of PID, td = 0",
    };
    
    static Cinfo vclampCinfo(
        "VClamp",
        Neutral::initCinfo(),
        vclampFinfos,
        sizeof(vclampFinfos) / sizeof(Finfo*),
        new Dinfo< VClamp >(),
        doc,
        sizeof(doc)/sizeof(string));

    return &vclampCinfo;            
}

static const Cinfo * vclampCinfo = VClamp::initCinfo();

VClamp::VClamp(): vIn_(0.0), command_(0.0), current_(0.0), mode_(0), ti_(0.0), td_(-1.0),
                  Kp_(0.0),
                  tau_(0.0),
                  tdByDt_(1.0),
                  dtByTi_(1.0),
                  e_(0.0),
                  e1_(0.0),
                  e2_(0.0)
{
}

VClamp::~VClamp()
{
    ;
}

void VClamp::setCommand(double value)
{
    // e2_ = 0;
    // e1_ = 0;
    cmdIn_ = value;
}

double VClamp::getCommand() const
{
    return command_;
}

void VClamp::setTi(double value)
{
    ti_ = value;
}

double VClamp::getTi() const
{
    return ti_;
}
void VClamp::setTd(double value)
{
    td_ = value;
}

double VClamp::getTd() const
{
    return td_;
}

void VClamp::setTau(double value)
{
    tau_ = value;
}

double VClamp::getTau() const
{
    return tau_;
}

void VClamp::setGain(double value)
{
    Kp_ = value;
}

double VClamp::getGain() const
{
    return Kp_;
}

double VClamp::getCurrent() const
{
    return current_;
}

void VClamp::setVin(double value)
{
    vIn_ = value;
}

void VClamp::setMode(unsigned int mode)
{
    mode_ = mode;
}

unsigned int VClamp::getMode() const
{
    return mode_;
}

void VClamp::process(const Eref& e, ProcPtr p)
{
    assert(ti_ > 0);
    assert(td_ >= 0);
    assert(tau_ > 0);
    double dCmd = cmdIn_ - oldCmdIn_;
    command_ = cmdIn_ + dCmd * ( 1 - tauByDt_) + (command_ - cmdIn_ + dCmd * tauByDt_) * expt_;
    oldCmdIn_ = cmdIn_;                           
    e_ = command_ - vIn_;
    if (mode_ == 0){
        current_ +=  Kp_ * ((1 + dtByTi_ + tdByDt_) * e_ - ( 1 + 2 * tdByDt_) * e1_ + tdByDt_ * e2_);
        e2_ = e1_;
        e1_ = e_;
    } else if (mode_ == DERIVATIVE_ON_PV){ // Here the derivative error term is replaced by process variable
        current_ +=  Kp_ * ((1 + dtByTi_) * e_ - e1_ + tdByDt_ * ( vIn_ - 2 * v1_ + e2_));
        e2_ = v1_;
        v1_ = vIn_;
        e1_ = e_;        
    } else if (mode_ == PROPORTIONAL_ON_PV){ // Here the proportional as well as the derivative error term is replaced by process variable
        current_ +=  Kp_ * (vIn_ - v1_ + dtByTi_ * e_ + tdByDt_ * ( vIn_ - 2 * v1_ + e2_));
        e2_ = v1_;
        v1_ = vIn_;
    }
    currentOut()->send(e, p->threadIndexInGroup, current_);
}

void VClamp::reinit(const Eref& e, ProcPtr p)
{

    vIn_ = 0.0;
    v1_ = 0;
    command_ = cmdIn_ = oldCmdIn_ = e_ = e1_ = e2_ = 0;
    if (ti_ == 0){
        ti_ = p->dt;
    }
    if (td_ < 0){
        td_ = 0.0;
    }
    if (tau_ == 0.0){
        tau_ = 5 * p->dt;
    }
    if (p->dt / tau_ > 1e-15){
        expt_ = exp(-p->dt/tau_);
    } else {
        expt_ = 1 - p->dt/tau_;
    }
    tauByDt_ = tau_ / p->dt;
    dtByTi_ = p->dt/ti_;
    tdByDt_ = td_ / p->dt;
    if (Kp_ == 0){
        vector<Id> compartments;
        unsigned int numComp = e.element()->getNeighbours(compartments, currentOut());
        if (numComp > 0){
            double Cm = Field<double>::get(compartments[0], "Cm");
            Kp_ = Cm / p->dt;
        }
    }
}

// 
// VClamp.cpp ends here

File: Demos/snippets/vclamp.py
# vclamp.py --- 
# 
# Filename: vclamp.py
# Description: 
# Author: 
# Maintainer: 
# Created: Sat Feb  2 19:16:54 2013 (+0530)
# Version: 
# Last-Updated: Mon Feb  4 19:14:45 2013 (+0530)
#           By: subha
#     Update #: 172
# URL: 
# Keywords: 
# Compatibility: 
# 
# 

# Commentary: 
# 
# 
# 
# 

# Change log:
# 
# 
# 
# 
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 3, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING.  If not, write to
# the Free Software Foundation, Inc., 51 Franklin Street, Fifth
# Floor, Boston, MA 02110-1301, USA.
# 
# 

# Code:

import sys
sys.path.append('/home/subha/src/moose/python')
import moose
sys.path.append('../squid')
from squid import SquidAxon
from pylab import *

def vclamp_demo(simtime=1000.0, dt=1e-2):
    # It is good practice to modularize test elements inside a
    # container
    container = moose.Neutral('/vClampDemo')
    clamp = moose.VClamp('/vClampDemo/vclamp')
    # Setup command voltage time course
    command = moose.PulseGen('/vClampDemo/command')
    command.delay[0] = 200.0
    command.width[0] = 100.0
    command.level[0] = -10.0
    command.delay[1] = 500.0
    command.level[1] = 55.0
    command.width[1] = 100.0
    moose.connect(command, 'outputOut', clamp, 'set_holdingPotential')
    # Create a compartment with properties of a squid giant axon
    comp = SquidAxon('/vClampDemo/axon')
    # Connect the Voltage Clamp to the compartemnt
    moose.connect(clamp, 'currentOut', comp, 'injectMsg')
    moose.connect(comp, 'VmOut', clamp, 'voltageIn')
    clamp.gain = comp.Cm/dt # This is a decent gain value
    # # This is for checking a current clamp
    # stim = moose.PulseGen('/vClampDemo/stim')
    # stim.delay[0] = 1e9
    # stim.width[0] = 100.0
    # stim.level[0] = 0.1
    # stim.delay[1] = 500.0
    # stim.level[1] = 0.2
    # stim.width[1] = 1e9
    # moose.connect(stim, 'outputOut', comp, 'injectMsg')
    # setup stimulus recroding
    # stimtab = moose.Table('/vClampDemo/stimtab')
    # moose.connect(stimtab, 'requestData', stim, 'get_output')
    # Set up Vm recording
    vmtab = moose.Table('/vClampDemo/vclamp_Vm')
    moose.connect(vmtab, 'requestData', comp, 'get_Vm')
    # setup command potential recording
    commandtab = moose.Table('/vClampDemo/vclamp_command')
    moose.connect(commandtab, 'requestData', clamp, 'get_holdingPotential')
    # setup current recording
    Imtab = moose.Table('/vClampDemo/vclamp_Im')
    moose.connect(Imtab, 'requestData', clamp, 'get_current')
    # Scheduling
    moose.setClock(0, dt)
    moose.setClock(1, dt)    
    moose.setClock(2, dt)
    moose.setClock(3, dt)
    moose.useClock(0, '%s/##[TYPE=Compartment]' % (container.path), 'init')
    moose.useClock(0, '%s/##[TYPE=PulseGen]' % (container.path), 'process')
    moose.useClock(1, '%s/##[TYPE=Compartment]' % (container.path), 'process')
    moose.useClock(2, '%s/##[TYPE=HHChannel]' % (container.path), 'process')
    moose.useClock(2, '%s/##[TYPE=VClamp]' % (container.path), 'process')
    moose.useClock(3, '%s/##[TYPE=Table]' % (container.path), 'process')
    moose.reinit()
    moose.start(simtime)
    print 'Finished simulation for %g seconds' % (simtime)
    tseries = linspace(0, simtime, len(vmtab.vec))
    subplot(211)
    title('Membrane potential and clamp voltage')
    plot(tseries, vmtab.vec, label='Vm (mV)')
    plot(tseries, commandtab.vec, label='Command (mV)')
    legend()
    # print len(commandtab.vec)
    subplot(212)
    title('Current through clamp circuit')
    # plot(tseries, stimtab.vec, label='stimulus (uA)')
    plot(tseries, Imtab.vec, label='Im (uA)')
    legend()
    show()

if __name__ == '__main__':
    vclamp_demo()
    

# 
# vclamp.py ends here
